Proposal: add cookies.removeAll() method

aselya opened the pull request:
Proposal introducing a new API cookies.removeAll()

aselya committed:
Create Proposal: add cookies.removeAll() method.md

Initial draft

aselya commented on proposals/Proposal: add cookies.removeAll() method.md:
````diff
@@ -0,0 +1,147 @@
+#### Special cases:
+```cookies.removeAll({})```
+
+>An empty details object will result in all cookies being removed. 
````
I'd like some input on this. My thought was to keep the behavior the same between `cookies.getAll` and this implementation but it might make sense to block it or direct developers to use the already existing browserData API

oliverdunk replied:
Open to thoughts, but I'd prefer to make this an error. As you mentioned, there is already the browsingData API for this, and even just during development it would be frustrating to accidentally clear all cookies. This would also allow us to match the CookieDetails object from the remove() API where `name` and `url` are currently required.

xeenon replied:
I would also like you make `{}` be an error. In Safari, we only return cookies you have host permissions for in `getAll({})`, so that is what would be removed if we supported it with `removeAll({})`. However, it does seem to be too much of a foot gun.

I think requiring one key is still useful and safe.

aselya replied:
Thank you for the feedback @xeenon and @oliverdunk , having `{}` be an error works for me.

I'm inclined to agree with @xeenon, that having at least one key is useful and safe. Would requiring the caller to provide at least a `url` or a `topLevelSite` sound good?

A `url` allows for a similar requirement as the existing `cookies.remove()` and it's inclusion allows for preserving the functionality. While a `topLevelSite` enables the new functionality of removing all cookies associated with a single `topLevelSite` introduced by the API. Keeping both of these parameters optional but requiring at least one of them, matches existing APIs such as [webNavigation.getFrame()](https://developer.chrome.com/docs/extensions/reference/api/webNavigation#method-getFrame) that already have the pattern of having all optional parameters but returning an error when no parameters or an invalid combination of parameters are provided.

oliverdunk replied:
I'm comfortable with that. Thanks!

oliverdunk commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,147 @@
+### Behavior
+
+The API will remove all cookies that match the `details` object parameter with the exception of the cases outlined in the implementation details.
+If the extension does not have  [host permissions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions)  for this URL, the API call will fail.
```
If we allow calling this API without a `url` (see below) we should add a note here about the behavior if you only have some host permissions. Presumably, we would just remove the cookies for the sites you have access to.

aselya replied:
Updated the API to require a `url`

oliverdunk reviewed (changes requested):
This generally looks good, thanks! I left a couple of comments.

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,147 @@
+
+## Motivation
+
+Prior to the introduction of [partitioned cookies](https://github.com/privacycg/CHIPS), the required parameters (`url` and `name`) used by the existing `cookies.remove()` api would be unique to a single cookie. The inclusion of an optional `partitionKey` means that there can be more than one cookie that matches the currently required parameters for `cookies.remove()`. Because of this, handling pa ...
```
> Prior to the introduction of [partitioned cookies](https://github.com/privacycg/CHIPS), the required parameters (`url` and `name`) used by the existing `cookies.remove()` api would be unique to a single cookie

This claim is not true. It was already possible for a given URL + name to potentially match more than one cookie. For context, see https://issues.chromium.org/issues/40572551 and https://bugzilla.mozilla.org/show_bug.cgi?id=1387957#c9

aselya replied:
Thank you for the historical context. I updated language in the motivation to be less `partitionKey` specific to reflect this in addition to the update to the related issues.

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,147 @@
+
+**Created:**  2024-09-18
+
+**Related Issues:**  [PrivacyCG/CHIPS issue 6 - How do Partitioned cookies interact with browser extensions?](https://github.com/privacycg/CHIPS/issues/6)
```
I suggest to also include https://issues.chromium.org/issues/40572551 and https://bugzilla.mozilla.org/show_bug.cgi?id=1387957

aselya replied:
Updated the related issues

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,147 @@
+
+### Objective
+
+This new API addresses two workflows for developers:
```
It seems that the examples you provided is how extensions would have to achieve the functionality these days. Can you rephrase to emphasize that? To someone not deeply aware of the details, it can be read as part of the proposal for something net, not of existing behavior.

aselya replied:
Updated the language and I would appreciate feedback on if my rephrasing addresses the issue you brought up.

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
````diff
@@ -0,0 +1,147 @@
+)
+```
+##### Parameters
+An `object` containing information to identify the cookie(s) to remove. It contains the following properties:
````
```suggestion
An `object` containing information to identify the cookie(s) to remove. It contains the following properties, based on the properties recognized by `cookies.remove`:
```

Added more context, to offer the reasoning for the choice of parameters. Note that the `remove` method does not contain enough properties to uniquely describe one cookie, but since `cookies.removeAll` may match multiple by design, I wouldn't expansion of the property set to be a hard requirement for this proposal.

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,147 @@
+>[`url`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#url) optional:
+>
+>A  `string`  representing the URL associated with the cookie.
+
```
"Return value" is not  documented in this proposal. What are you going to return?

The `remove` method returns the deleted cookie, and that could be reasonable to do the same in `cookies.removeAll`. A concrete use case is to allow for extensions to undo the removal. Currently extensions have to call `cookies.get` or `cookies.getAll` and keep track of the cookie. This tracking requires extensions to accurately predict which cookie would be removed for a given `cookies.removeAll` call, but it is not easy to determine which cookies would be removed.

Another option is to return nothing (undefined) and encourage extension developers to call `getAll` with the given parameters, under the assumption that the `cookies.getAll` call would return exactly the same set of cookies as matched by `cookies.removeAll`. In theory this is prone to TOCTOU issues, but I think that this is not that big of a deal in practice.

Yet another option is to add an option to the `cookies.removeCookies` call to flag whether deleted cookies should be returned.

aselya replied:
I think returning a list of cookies that were removed makes sense here. In addition to addressing the use case you mentioned it also matches the behavior of `cookies.getAll()`.

aselya committed:
Update proposals/Proposal: add cookies.removeAll() method.md

Updating parameters to explicitly reference cookies.remove

Co-authored-by: Rob Wu <rob@robwu.nl>

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Add more related issues
Update the motivation language to reflect the related issues.

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Add link to bug in the motivation

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Update the objective to be more explicit about the pain point for developers.

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Update language in objective.

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Add return

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Update language to make `{}` and error

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
````diff
@@ -0,0 +1,157 @@
+    partitionKey:{}})
+```
+
+>An empty `partitionKey` object as the only value in the details object, will result in all partitioned cookies being removed.
````
`partitionKey: {}` in `cookies.getAll` matches all cookies, partitioned and unpartitioned: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/getAll#partitionkey

Can we match that behavior, and if not, why do we deviate from it?

aselya replied:
My intention was to allow a caller to easily choose between removing just partitioned cookies  or just unpartitioned cookies (by not including a partitionKey). But it keeping the semantics the same between this API and `cookies.getAll` is better for developers as it's one less quirk for them to remember.

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
````diff
@@ -0,0 +1,157 @@
+
+```
+cookies.removeAll({
+    topLevelSite: “https://example.com”
````
Why is topLevelSite a separate key? I would expect it to be a member of partitionKey.

aselya replied:
Having `topLevelSite` be a separate (optional) parameter addresses the second motivation "Removal of cookies associated with a topLevelSite". 

This could be replaced by having the developer pass in a partitionKey that contains the `toplevelSite` and no value for `hasCrossSiteAncestor`, such as`partitionKey:{topLevelSite: "https/example.com"}`.  But I wanted to provide an easier option for developers who may not be as familiar with the intricacies of partitioned cookies.

aselya replied:
Based on feedback, I updated the api design, to no longer have `topLevelSite` as a separate key.

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Update case: `partitionKey : {}` to match `cookies.getAll`

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Update requirements of details object.

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Added more detailed description of host permissions impact on API call

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Made update to remove the `top-level site` as an optional stand alone parameter.

oliverdunk reviewed (approved):
This looks good to me! Adding @rdcronin to make sure he is happy before we land.

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,155 @@
+>
+>A  `string`  representing the ID of the cookie store to find the cookie in. If unspecified, the cookie is looked for by default in the current execution context's cookie store.
+>
+>[`url`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#url)
```
```suggestion
>[`url`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/remove#url) optional
```

The sentence two lines below implies that `url` can be optional.

Rob--W commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,155 @@
+
+### Behavior
+
+The API will remove all cookies that match the `details` object parameter with the exception of the cases outlined in the implementation details.
```
"all cookies that match" is doing some heavy lifting. Would it be meaningful to refer to `cookies.getAll` as the primitive? Or are there differences?

aselya replied:
Made update to language that references `cookies.getAll`

Rob--W reviewed (approved):
Some examples with non-trivial cases would be nice, but

rdcronin commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,155 @@
+>
+>A  `string`  representing the URL associated with the cookie.
+>
+>A details object must contain a `url` and/or a `topLevelSite`.
```
nit: topLevelSite is not a (direct) property here, so maybe tweak to:

```
>A details object must contain at least a `url` a `partitionKey` with a valid `topLevelSite`.
```

aselya replied:
Done

rdcronin commented on proposals/Proposal: add cookies.removeAll() method.md:
```diff
@@ -0,0 +1,155 @@
+
+##### Return
+
+> `Promise<Cookie[]>`:
```
/cc @Rob--W since I know he normally has thoughts on performance-related aspects.

The existing remove() method doesn't return a full cookie, but rather a subset of its properties: name, partitionKey, storeId, and url.  I'm curious if we should perhaps lean on that instead, which avoids serializing quite as much data -- in particular, it avoids passing all the values to the extension (which may be quite large in practice).

WDYT?

(If we go that route, we could introduce a common type for it -- `CookieSummary`, `CookieMetadata`, or similar?)

rdcronin reviewed (approved):
Thanks, @aselya !  Generally LG, but a few comments.

aselya committed:
Update proposals/Proposal: add cookies.removeAll() method.md

Co-authored-by: Rob Wu <rob@robwu.nl>

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Update to add reference to `cookies.getAll` as the matching behavior primitive.

aselya committed:
Update Proposal: add cookies.removeAll() method.md

Update language to - A details object must contain at least a `url` a `partitionKey` with a valid `topLevelSite`.
